File: services/user-service/src/main/java/com/visor/school/userservice/model/UserPermission.java
package com.visor.school.userservice.model;

import java.time.Instant;
import java.util.UUID;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Index;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;

@Entity
@Table(name = "user_permissions", indexes = {
    @Index(name = "idx_user_permissions_user_id", columnList = "user_id"),
    @Index(name = "idx_user_permissions_permission_id", columnList = "permission_id"),
    @Index(name = "idx_user_permissions_unique", columnList = "user_id,permission_id", unique = true)
})
public class UserPermission {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "permission_id", nullable = false)
    private Permission permission;

    @Column(name = "assigned_at", nullable = false)
    private Instant assignedAt = Instant.now();

    @Column(name = "assigned_by")
    private UUID assignedBy;

    protected UserPermission() {}

    public UserPermission(User user, Permission permission, UUID assignedBy) {
        this.user = user;
        this.permission = permission;
        this.assignedBy = assignedBy;
    }

    public UUID getId() {
        return id;
    }

    public User getUser() {
        return user;
    }

    public Permission getPermission() {
        return permission;
    }

    public Instant getAssignedAt() {
        return assignedAt;
    }

    public UUID getAssignedBy() {
        return assignedBy;
    }
}

File: services/user-service/src/main/kotlin/com/visor/school/userservice/model/UserPermission.kt
package com.visor.school.userservice.model

import jakarta.persistence.*
import java.time.Instant
import java.util.UUID

/**
 * Junction entity linking users to permissions
 */
@Entity
@Table(name = "user_permissions", indexes = [
    Index(name = "idx_user_permissions_user_id", columnList = "user_id"),
    Index(name = "idx_user_permissions_permission_id", columnList = "permission_id"),
    Index(name = "idx_user_permissions_unique", columnList = "user_id,permission_id", unique = true)
])
class UserPermission(
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    val id: UUID = UUID.randomUUID(),

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    val user: User,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "permission_id", nullable = false)
    val permission: Permission,

    @Column(name = "assigned_at", nullable = false)
    val assignedAt: Instant = Instant.now(),

    @Column(name = "assigned_by")
    val assignedBy: UUID? = null
)

